### DISCUSSION TITLE
Token registry integrations

### DISCUSSION THREAD

**Discussion by marlowl** [Open-Attestation/open-attestation#240] [Q&A] [answered]
@superical @simboonlong  I checked out the beta branch of the token-registry, looks like quite an big upgrade compared to the v3 versions. Is there any timeline of when this beta will be released into a 'full' major v4 release?

And which of the versions would you recommend to build stable integrations against? I could imagine this is still the v2 packages, since these versions are also integrated, in other TradeTrust/OA tooling.  (see [this](https://github.com/Open-Attestation/open-attestation-cli/blob/master/package.json#L78) and [this](https://github.com/TradeTrust/tradetrust-website/blob/master/package.json#L52) for example).

**Comment by superical** [comment, marked as answer]
hey @marlowl! 

>  Is there any timeline of when this beta will be released into a 'full' major v4 release?

The v4 beta contracts are undergoing an audit. I'd like to think that a good time for transiting into a full major v4 could be when the audit is completed and issues are rectified. 

> And which of the versions would you recommend to build stable integrations against?

I'm not sure if there is an official recommendation from the product team about the version, but v3 is currently the "latest" distribution version. I'll have to check on this if you need.

But if you were asking for my _personal_ (hence unofficial😅) opinion from a technical perspective, I would go with the v4 beta as much as I can. Just my 2 sats though.

> I could imagine this is still the v2 packages, since these versions are also integrated, in other TradeTrust/OA tooling.

Yea, few of them are still stuck at some of the early versions of v2. There are works in the pipeline to update them.

**Reply by marlowl** [reply]
@superical 

>The v4 beta contracts are undergoing an audit. I'd like to think that a good time for transiting into a full major v4 could be when the audit is completed and issues are rectified.

>But if you were asking for my  _personal_  (hence unofficial) opinion from a technical perspective, I would go with the v4 beta as much as I can. Just my 2 sats though.

Clear, pretty cool to hear that they are being audited!
Until the full release, I will take into consideration the  'be warned, but be brave' warning in the docs. :stuck_out_tongue_winking_eye:

>I'm not sure if there is an official recommendation from the product team about the version, but v3 is currently the "latest" distribution version. I'll have to check on this if you need.

> Yea, few of them are still stuck at some of the early versions of v2. There are works in the pipeline to update them.

From a pure implementer  perspective, this would definitely be useful. (Since I noticed that there are also some major differences between v2 and v3). That's why I looked for reference at the current implemented frontend package versions in the OA/TradeTrust tooling.

**Reply by superical** [reply]
Yup, there are currently works in the pipeline to update them. 

My teammate @puayhiang is working on updating the frontend and TT toolings such as the CLI et al to use the latest version. The CLI is currently in the midst of it. You can check the status in the PR https://github.com/Open-Attestation/open-attestation-cli/pull/212 if you are interested. If that PR gets passed, the CLI should be updated to use the v4 of token registry.😊

**Reply by marlowl** [reply]
@superical Perfect, will definitely take a look at the v4 implementations

Thanks for the update!

**Reply by marlowl** [reply]
@superical Did have some time to look at the v4 implementations last days. My opinion/feedback on it:
- Contracts are looking solid and quite heavily tested to me. Out of curiosity, is it possible for you to name the auditing firm?
- Some of the hardhat readme examples were not working, for example [this](https://github.com/Open-Attestation/token-registry/tree/beta#using-an-existing-title-escrow-factory) one
- From a pure implementer perspective, the Frontend/Tooling/NPM packages around the new v4 contracts still look work-in-progress to me (as you already mentioned)

Also open to for example have a technical alignment/timeline call, if you think that would be more effective! :smile:

**Reply by superical** [reply]
> Contracts are looking solid and quite heavily tested to me. Out of curiosity, is it possible for you to name the auditing firm?

Yea, it was one of our goals to make sure that v4 is more thoroughly tested compared to the previous versions. As for the audit, the devs didn't get to decide the auditing firm, but it was decided for us that EY will be auditing the contracts.

> Some of the hardhat readme examples were not working, for example [this](https://github.com/Open-Attestation/token-registry/tree/beta#using-an-existing-title-escrow-factory) one

Ohh.. Would you mind elaborating what's not working when deploying with an existing title escrow? Are you getting any errors and any specific steps to reproduce your issue? I can have a look at it.

> From a pure implementer perspective, the Frontend/Tooling/NPM packages around the new v4 contracts still look work-in-progress to me (as you already mentioned)

Yea, the update to the newer version is still work in progress.

**Reply by marlowl** [reply]
@superical This is one of the repro cases I can build:

**Deploy a factory**
```sh
npx hardhat deploy:factory --network goerli
[Deployer] 0x0861152F372688F61112ece01B2a8Af9BE06e0a1
[Transaction] Pending 0x52679c010fa1a5499a2a57eb81f69658fe124597b395f82ac5df6410985a06b2
[Address] Deployed to 0x762fc6bF4Ac40e49677E9a577F9c4d3e46D1D3eE
[Transaction] Pending 0x52679c010fa1a5499a2a57eb81f69658fe124597b395f82ac5df6410985a06b2
[Status] Deployed to 0x762fc6bF4Ac40e49677E9a577F9c4d3e46D1D3eE
[Status] ✅ Completed deploying Title Escrow Factory at 0x762fc6bF4Ac40e49677E9a577F9c4d3e46D1D3eE
```

**Trying to deploy a token to a given factory**
```sh
npx hardhat deploy:token --name ebl --symbol ebl --network goerli --factory 0x762fc6bF4Ac40e49677E9a577F9c4d3e46D1D3eE
[Deployer] 0x0861152F372688F61112ece01B2a8Af9BE06e0a1
[Status] Title Escrow Factory interface check is OK.
[Transaction] Pending 0x908221eb5852e08be0696ba1ad604b155cc35a44497f8cd95a452391c5778e90
[Status] ❌ An error occurred while deploying token
Cannot find topic 0x3588ebb5c75fdf91927f8472318f41513ee567c2612a5ce52ac840dcf6f162f5
```

**Reply by superical** [reply]
Ahh... This looks like a bug in the address config + goerli isn't fully supported yet for quick start mode in token registry. As mentioned in the [readme](https://github.com/Open-Attestation/token-registry/tree/beta#quick-start), only mumbai, rinkeby, and ropsten are currently supported for quick start. 

Coincidentally, I'm right in the midst of pushing support for goerli and sepolia for the token registry after merging the same support for the frontend and cli this week.😄 I should be able to add support for goerli in token registry ready by today or early tomorrow. I have also pushed in a fix for the address bug in https://github.com/Open-Attestation/token-registry/pull/126.

**Reply by superical** [reply]
@marlowl The update for goerli and sepolia in quick start mode has been merged into the beta branch yesterday. Let us know if it works for you now. 😊

**Reply by marlowl** [reply]
@superical can confirm it works on my side as well now! 😄

**Reply by superical** [reply]
Nice! Glad that it works for you now!😃

**Comment by marlowl** [comment]
@superical @simboonlong  I noticed that the major [v4](https://github.com/Open-Attestation/token-registry/releases/tag/v4.0.0) version of the token-registry was released, nice 😄  

Does the v4 version already contain the audited version of the smart contracts?

**Reply by superical** [reply]
Still working out a couple of minor items to be confirmed with the auditor though.

**Reply by marlowl** [reply]
@superical Thanks for clarifying!
Once those minor items are cleared, I guess there will be something like another major release?

**Reply by superical** [reply]
I don't think so. We promoted the version from beta to major only because product team requested it for their publication and so there are still some remaining changes to be confirmed. I'm not expecting the changes to be too far from what it is now though. Perhaps I can update this thread when the remaining items are all confirmed so that you will be aware of it?

**Reply by marlowl** [reply]
That would be great, thanks! 😄
